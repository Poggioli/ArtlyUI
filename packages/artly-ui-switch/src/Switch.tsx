import { forwardRef } from "@artly-ui/core";
import { useContext, useEffect } from "react";
import { SwitchContext, SwitchContextProvider } from "./context";
import {
  StyledSwitchContainer,
  StyledSwitchLabel,
  StyledSwitchRoot,
  StyledSwitchThumb,
} from "./styles";
import { SwitchContainerProps, SwitchLabelProps, SwitchProps } from "./types";

// ------------------------------------------- Switch Container ------------------------------------------- //

const SwitchContainer = forwardRef<
  typeof StyledSwitchContainer,
  SwitchContainerProps
>(({ children, ...props }, forwardedRef) => {
  return (
    <SwitchContextProvider>
      <StyledSwitchContainer
        {...props}
        ref={forwardedRef}
        align="center"
        justify="space-between"
        gap="$3"
        wrap="nowrap"
      >
        {children}
      </StyledSwitchContainer>
    </SwitchContextProvider>
  );
});

// ------------------------------------------- Switch Label ------------------------------------------- //

const SwitchLabel = forwardRef<typeof StyledSwitchLabel, SwitchLabelProps>(
  ({ ...props }, forwardedRef) => {
    const {
      state: { disabled },
    } = useContext(SwitchContext);

    return (
      <StyledSwitchLabel
        {...props}
        ref={forwardedRef}
        variant="body1"
        color={disabled ? "$gray-80" : props.color}
      />
    );
  }
);

// ------------------------------------------- Switch Root ------------------------------------------- //

const Switch = forwardRef<typeof StyledSwitchRoot, SwitchProps>(
  ({ disabled, color, ...props }, forwardedRef) => {
    const { setState } = useContext(SwitchContext);

    useEffect(() => {
      setState((previousState) => ({
        ...previousState,
        disabled: !!disabled,
      }));
    }, [disabled, setState]);

    return (
      <StyledSwitchRoot
        {...props}
        ref={forwardedRef}
        disabled={disabled}
        color={color}
      >
        <StyledSwitchThumb color={color} />
      </StyledSwitchRoot>
    );
  }
);

export { Switch, SwitchContainer, SwitchLabel };
