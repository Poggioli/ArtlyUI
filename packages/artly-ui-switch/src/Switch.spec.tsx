import { render } from "@testing-library/react";
import { Switch } from "./Switch";

describe("Core component / Switch", () => {
  it(`GIVEN a Switch component
    WHEN render component
    SHOULD occurs render normaly`, () => {
    const rendered = render(<Switch />);
    expect(rendered.container).toBeTruthy();
  });
});
