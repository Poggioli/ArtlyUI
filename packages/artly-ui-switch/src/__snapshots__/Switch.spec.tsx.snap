// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core component / Switch GIVEN a Switch component
      WITH disabled state
      WHEN render component
      THEN should occurs render normaly 1`] = `
<div>
  <div
    class="artly-UI-c-bXbWpx artly-UI-c-bXbWpx-igIyMSm-css artly-UI-c-PJLV"
  >
    <label
      class="artly-UI-c-chWhuJ artly-UI-c-chWhuJ-jprvTt-variant-body1 artly-UI-c-chWhuJ-cmVlgk-align-left artly-UI-c-chWhuJ-ijDMiJS-css artly-UI-c-kwAZbD"
      for="switch-airplane"
    >
      Switch Label
    </label>
    <button
      aria-checked="false"
      class="artly-UI-c-ctpKDW artly-UI-c-ctpKDW-dRUqVD-color-primary"
      data-disabled=""
      data-state="unchecked"
      disabled=""
      id="switch-airplane"
      role="switch"
      type="button"
      value="on"
    >
      <span
        class="artly-UI-c-ihrdaH artly-UI-c-ihrdaH-enlfht-color-primary"
        data-disabled=""
        data-state="unchecked"
      />
    </button>
  </div>
</div>
`;
